storagenode/piecestore/retain.go

```diff
@@ -0,0 +1,186 @@
+// See LICENSE for copying information.
+
+package piecestore
+
+import (
+	"context"
+	"runtime"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+
+	"storj.io/storj/pkg/bloomfilter"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/storagenode/pieces"
+)
+
+// RetainService queues and processes retain requests from satellites
+type RetainService struct {
+	log          *zap.Logger
+	retainStatus RetainStatus
+
+	mu     sync.Mutex
+	queued map[storj.NodeID]RetainRequest
+
+	reqChan      chan RetainRequest
+	sem          chan struct{}
+	emptyTrigger chan struct{}
+
+	store *pieces.Store
+}
+
+// RetainRequest contains all the info necessary to process a retain request
+type RetainRequest struct {
+	SatelliteID   storj.NodeID
+	CreatedBefore time.Time
+	Filter        *bloomfilter.Filter
+}
+
+// NewRetainService creates a new retain service
+func NewRetainService(log *zap.Logger, retainStatus RetainStatus, concurrentRetain int, store *pieces.Store) *RetainService {
+	return &RetainService{
+		log:          log,
+		retainStatus: retainStatus,
+		queued:       make(map[storj.NodeID]RetainRequest),
+		reqChan:      make(chan RetainRequest),
+		sem:          make(chan struct{}, concurrentRetain),
+		emptyTrigger: make(chan struct{}),
+		store:        store,
+	}
+}
+
+// QueueRetain adds a retain request to the queue
+func (s *RetainService) QueueRetain(req RetainRequest) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// only queue retain request if we do not already have one for this satellite
+	if _, ok := s.queued[req.SatelliteID]; !ok {
+		s.queued[req.SatelliteID] = req
+		go func() { s.reqChan <- req }()
+	}
+}
+
+// Run listens for queued retain requests and processes them as they come in
+func (s *RetainService) Run(ctx context.Context) error {
+	for {
+		// exit if context has been canceled. Otherwise, block until an item can be added to the semaphore
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case s.sem <- struct{}{}:
+		}
+
+		// get the next request
+		var req RetainRequest
+		select {
+		case req = <-s.reqChan:
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+
+		// TODO make it possible to sync with this goroutine
+		go func(ctx context.Context, req RetainRequest) {
+			err := s.retainPieces(ctx, req)
+			if err != nil {
+				s.log.Error("retain error", zap.Error(err))
+			}
+			s.mu.Lock()
+			delete(s.queued, req.SatelliteID)
+			s.mu.Unlock()
+
+			if len(s.queued) == 0 {
+				s.emptyTrigger <- struct{}{}
+			}
+
+			// remove item from semaphore and free up process for another retain job
+			<-s.sem
+		}(ctx, req)
+	}
+}
+
+// Wait blocks until the context is canceled or until the queue is empty
+func (s *RetainService) Wait(ctx context.Context) {
+	s.mu.Lock()
+	queueLength := len(s.queued)
+	s.mu.Unlock()
+	if queueLength == 0 {
+		return
+	}
+	select {
+	case <-s.emptyTrigger:
+	case <-ctx.Done():
+	}
+}
+
+func (s *RetainService) retainPieces(ctx context.Context, req RetainRequest) (err error) {
+	defer mon.Task()(&ctx)(&err)
```

Comment:
We shouldn't have to worry about anything being queued if retain status is disabled (there is another check in `endpoint.Retain` before queue is called). If you still think I should make the change, I can.

Reply:
I did update the monkit call to have the parameters you mentioned.

Reply:
> We shouldn't have to worry about anything being queued if retain status is disabled (there is another check in endpoint.Retain before queue is called). If you still think I should make the change, I can

I don't fully understand this comment because I think that I didn't explain myself clearly.

My worries are the following:

1. If we call `retainPieces` when enabled or debug then this function call will be shown in monkit as a function that it took some time to performs.
1. If we call `retainPieces` when disabled then this function call will be shown in monkit as a function that ti took very minimal time to performs.

If at some point we would like to analyze some data monkit data collected from several Storage Nodes and we want to see the time that they spend calling `retainPieces` we are going to have some mixed data which between the ones which are enabled/debug and disabled but we are not going to have a sway to exclude the disabled ones because the status isn't informed in the parameters.

Because I think that it's worthless to track calls when disabled, I was suggesting not to perform the call rather than adding the status parameter.

I'm expecting you change it if my point doesn't make sense considering that I'm not sure how we analyze the monkit data and the exposed case isn't real.

Reply:
What I am saying is that in `piecestore.endpoint.Retain`, we do not call `retain.Queue()` if retain status is set to disabled. Which means that in production, `retainPieces` will never be called if retain status is disabled, even if we do not explicitly check for it here. But that might not always be true in the future so I can add the condition just in case.

Reply:
Ah, OK, I do understand now.
I thought that it was called due to the condition that there is just in the lines below this one.

If we actually check for it just in case, shouldn't we move this call after that conditional?

Reply:
yes, I updated it